src/core/net/dnssd_server.cpp

```diff
@@ -839,6 +855,121 @@ Error Server::ResolveByQueryCallbacks(Header                 &aResponseHeader,
     return error;
 }
 
+#if OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE
+bool Server::ShouldForwardToUpstream(const Header &aRequestHeader, const Message &aRequestMessage)
+{
+    bool     ret = true;
+    uint16_t readOffset;
+    char     name[Name::kMaxNameSize];
+
+    VerifyOrExit(aRequestHeader.IsRecursionDesiredFlagSet(), ret = false);
+    readOffset = sizeof(Header);
+
+    for (uint16_t i = 0; i < aRequestHeader.GetQuestionCount(); i++)
+    {
+        VerifyOrExit(kErrorNone == Name::ReadName(aRequestMessage, readOffset, name, sizeof(name)), ret = false);
+        readOffset += sizeof(Question);
+
+        VerifyOrExit(!Name::IsSubDomainOf(name, kDefaultDomainName), ret = false);
+        for (const char *blockedDomain : kBlockedDomain)
+        {
+            VerifyOrExit(!Name::IsSameDomain(name, blockedDomain), ret = false);
+        }
+    }
+
+exit:
+    return ret;
+}
+
+void Server::OnUpstreamQueryResponse(UpstreamQueryTransaction &aQueryTransaction, Message &aResponseMessage)
+{
+    Error err = kErrorNone;
+
+    VerifyOrExit(aQueryTransaction.IsValid());
+
+    err = mSocket.SendTo(aResponseMessage, aQueryTransaction.GetMessageInfo());
+    if (err != kErrorNone)
+    {
+        aResponseMessage.Free();
+    }
+
+    aQueryTransaction.Reset();
+    ResetTimer();
+
+exit:
+    return;
+}
+
+Server::UpstreamQueryTransaction *Server::AllocateUpstreamQueryTransaction(const Header           &aRequestHeader,
+                                                                           const Ip6::MessageInfo &aMessageInfo)
+{
+    UpstreamQueryTransaction *ret = nullptr;
+
+    for (UpstreamQueryTransaction &txn : mUpstreamQueryTransactions)
+    {
+        if (!txn.IsValid())
+        {
+            ret = &txn;
+            txn.Init(aRequestHeader, aMessageInfo);
+            LogInfo("Allocated new transaction for upstream query.");
+            break;
+        }
+    }
+
+    if (ret != nullptr)
+    {
+        mTimer.FireAtIfEarlier(ret->GetExpireTime());
+    }
+
+    return ret;
+}
+
+Error Server::ResolveByUpstream(const Header           &aRequestHeader,
+                                Message                &aRequestMessage,
+                                const Ip6::MessageInfo &aMessageInfo)
+{
+    Error                     error = kErrorNone;
+    UpstreamQueryTransaction *txn   = nullptr;
+
+    txn = AllocateUpstreamQueryTransaction(aRequestHeader, aMessageInfo);
+    VerifyOrExit(txn != nullptr, error = kErrorNoBufs);
+
+    LogInfo("Forwarding DNS query to upstream");
+    otPlatDnsStartUpstreamQuery(&GetInstance(), txn, &aRequestMessage);
```

Comment:
Same here about logging.
We log at `AllocateUpstreamQueryTransaction()` and then again here (and I think these are the only new logs for "upstream" feature).
```
Allocated new transaction for upstream query.
Forwarding DNS query to upstream
```